src/common/procedure/src/procedure.rs

```diff
@@ -0,0 +1,239 @@
+// Copyright 2023 Greptime Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::fmt;
+use std::sync::Arc;
+
+use async_trait::async_trait;
+use serde::{Deserialize, Serialize};
+use uuid::Uuid;
+
+use crate::error::Result;
+
+/// Procedure execution status.
+pub enum Status {
+    /// The procedure is still executing.
+    Executing {
+        /// Whether the framework needs to persist the procedure.
+        persist: bool,
+    },
+    /// The procedure has suspended itself and is waiting for subprocedures.
+    Suspended {
+        subprocedures: Vec<ProcedureWithId>,
+        /// Whether the framework needs to persist the procedure.
+        persist: bool,
+    },
+    /// the procedure is done.
+    Done,
+}
+
+impl Status {
+    /// Returns a [Status::Executing] with given `persist` flag.
+    pub fn executing(persist: bool) -> Status {
+        Status::Executing { persist }
+    }
+
+    /// Returns `true` if the procedure needs the framework to persist its intermediate state.
+    pub fn need_persist(&self) -> bool {
+        // If the procedure is done, the framework doesn't need to persist the procedure
+        // anymore. It only needs to mark the procedure as committed.
+        match self {
+            Status::Executing { persist } | Status::Suspended { persist, .. } => *persist,
+            Status::Done => false,
+        }
+    }
+}
+
+/// Procedure execution context.
+#[derive(Debug)]
+pub struct Context {
+    /// Id of the procedure.
+    pub procedure_id: ProcedureId,
+}
+
+/// A `Procedure` represents an operation or a set of operations to be performed step-by-step.
+#[async_trait]
+pub trait Procedure: Send + Sync {
+    /// Type name of the procedure.
+    fn type_name(&self) -> &str;
+
+    /// Execute the procedure.
+    ///
+    /// The implementation must be idempotent.
+    async fn execute(&mut self, ctx: &Context) -> Result<Status>;
+
+    /// Dump the state of the procedure to a string.
+    fn dump(&self) -> Result<String>;
```

Comment:
Maybe we can just make `Procedure` trait extend `Display` trait?

Reply:
This method behaves like `serialize()` and is different from `display`